tket/src/WeightSubgrMono/Searching/NodeListTraversal.cpp

```diff
@@ -0,0 +1,289 @@
+// Copyright 2019-2022 Cambridge Quantum Computing
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "WeightSubgrMono/Searching/NodeListTraversal.hpp"
+
+#include <sstream>
+
+#include "Utils/Assert.hpp"
+#include "WeightSubgrMono/Common/GeneralUtils.hpp"
+#include "WeightSubgrMono/Searching/NodesRawData.hpp"
+
+namespace tket {
+namespace WeightedSubgraphMonomorphism {
+
+NodeListTraversal::NodeListTraversal(NodesRawDataWrapper& raw_data_wrapper)
+    : m_raw_data(raw_data_wrapper.m_raw_data) {}
+
+std::set<VertexWSM> NodeListTraversal::get_used_target_vertices() const {
+  std::set<VertexWSM> target_vertices;
+
+  // Examine all PV.
+  for (unsigned pv = 0; pv < m_raw_data.domains_data.size(); ++pv) {
+    const auto& domain_data = m_raw_data.domains_data[pv];
+
+    // Examine Dom(PV) at all levels.
+    for (unsigned level = 0; level <= domain_data.entries_back_index; ++level) {
+      const auto& node_index = domain_data.entries[level].node_level;
+      // Nogood nodes are ignored.
+      // So, check that at least one node with this domain
+      // is not a nogood; otherwise skip.
+      if (m_raw_data.nodes_data[node_index].nogood &&
+          level < domain_data.entries_back_index) {
+        const auto& next_node_index = domain_data.entries[level + 1].node_level;
+        bool some_node_is_good = false;
+        for (auto ii = node_index; ii < next_node_index; ++ii) {
+          if (!m_raw_data.nodes_data[ii].nogood) {
+            some_node_is_good = true;
+            break;
+          }
+        }
+        if (!some_node_is_good) {
+          // Skip.
+          continue;
+        }
+      }
+      const auto& domain = domain_data.entries[level].domain;
+      if (target_vertices.empty()) {
+        target_vertices = domain;
+      } else {
+        for (VertexWSM tv : domain) {
+          target_vertices.insert(tv);
+        }
+      }
+    }
+  }
+  return target_vertices;
+}
+
+bool NodeListTraversal::move_up() {
+  while (m_raw_data.current_node_level > 0) {
+    --m_raw_data.current_node_level;
+    if (m_raw_data.nodes_data[m_raw_data.current_node_level].nogood) {
+      continue;
+    }
+    for (unsigned pv = 0; pv < m_raw_data.domains_data.size(); ++pv) {
+      auto& domain_data = m_raw_data.domains_data[pv];
```

Comment:
Sorry, no, this is incorrect.

Look in src\WeightSubgrMono\include\WeightSubgrMono\Searching\NodesRawData.hpp.

If `NodesRawData` has `j` `NodeData` objects in `nodes_data` then there are `j` nodes. _(Actually not quite, since `nodes_data` for performance reasons is allowed to have extra "junk" data - hence the variable `unsigned current_node_level;` so we know how many nodes are ACTUAL nodes, not junk to be cleared and overwritten later. But ignore this point - actually I will wrap this "logical" resizing in a new template to make it a bit easier)._

`domains_data` does include information, for every pattern vertex. BUT this information, which has a vector including `std::set<VertexWSM> domain` and other things, does NOT have size `j` necessarily. _(But you are right, that `std::set<VertexWSM> domain` exactly is the collection of target vertices which that particular pattern vertex still might get assigned to as we move further down the search)._

For performance, a new `std::set<VertexWSM> domain` object is stored (within the vector) only when it CHANGES; different domains (for different pv) change at different nodes.

This cuts down on the amount of copying. Otherwise we'd have to copy a potentially large `std::set<VertexWSM>` for EVERY pv, at EVERY node, even if almost all the data is unchanged from one node to the next.

I will add more comments explaining this.

Think of `nodes_data` as, LOGICALLY, a collection of horizontal "views", and `domains_data` as a LOGICAL collection of vertical "views", into THE EXACT SAME DATA, stored within the `NodesRawData` object. Then realise that, we don't need  data overlap; the data is spread between the two. Finally, realise that the vertical "views" can be "compressed" by removing redundant information _(i.e., domains which haven't changed since the previous node)._

Reply:
Anyway, the `auto` is gone!